egs/wsj/s5/steps/nnet3/train_rnn.py

```diff
@@ -0,0 +1,1134 @@
+    # use during decoding
+    CopyEgsPropertiesToExpDir(egs_dir, args.dir)
+
+    if (args.stage <= -3):
+        logger.info('Computing the preconditioning matrix for input features')
+        
+        ComputePreconditioningMatrix(args.dir, egs_dir, run_opts,
+                                     num_lda_jobs = num_archives,
+                                     max_lda_jobs = args.max_lda_jobs,
+                                     rand_prune = args.rand_prune)
+
+    if (args.stage <= -2):
+        logger.info("Preparing initial vector for FixedScaleComponent before softmax using priors^$presoftmax_prior_scale_power and rescaling to average 1")
+        ComputePresoftmaxPriorScale(dir, alidir, num_jobs, run_opts,
+                                    args.presoftmax_prior_scale_power)
+
+    if (args.stage <= -1):
+        logger.info("Preparing the initial acoustic model.")
+        PrepareInitialAcousticModel(dir, alidir, run_opts)
+
+    
+    # set num_iters so that as close as possible, we process the data $num_epochs
+    # times, i.e. $num_iters*$avg_num_jobs) == $num_epochs*$num_archives,
+    # where avg_num_jobs=(num_jobs_initial+num_jobs_final)/2.
+    num_archives_to_process= args.num_epochs * num_archives
+    num_iters=(num_archives_to_process * 2) / (args.num_jobs_initial + args.num_jobs_final)
+
+    num_iters_combine = VerifyIterations(num_iters, num_epochs,
+                                         num_hidden_layers, num_archives,
+                                         max_models_combine, add_layers_period,
+                                         num_jobs_final)
+
+    learning_rate = lambda x: GetLearningRate(x, num_iters,
+                                              num_archives_to_process,
+                                              args.initial_effective_lrate,
+                                              args.final_effective_lrate)
+
+    if args.realign_times is not None:
+        realign_iters = GetRealignIters(args.realign_times,
+                                        args.num_jobs_initial,
+                                        args.num_jobs_final)
+
+    # egs_dir will be updated if there is realignment
+    cur_egs_dir=egs_dir
+
+    if args.num_bptt_steps is None:
+        num_bptt_steps = args.chunk_width
+    else:
+        num_bptt_steps = args.num_bptt_steps
+
+    min_deriv_time = args.chunk_width - num_bptt_steps
+
+    for iter in range(num_iters):
+        if (args.exit_stage is not None) and (iter == args.exit_stage):
+            logger.info("Exiting early due to --exit-stage {0}".format(iter))
+            return
+        current_num_jobs = int(0.5 + args.num_jobs_initial + (args.num_jobs_final - args.num_jobs_initial) * iter / num_iters)
+        
+        if iter in realign_iters:
+            prev_egs_dir=cur_egs_dir
+            cur_egs_dir="{0}/egs_{1}".format(args.dir, "iter"+str(iter))
+            Realign(args.dir, iter, args.alidir, args.lang,
```

Comment:
this should probably be inside a check for 'stage'.  If it isn't in the baseline script it would be a bug.